samples/bluetooth/alexa_gadget/src/gadgets/CMakeLists.txt

```diff
@@ -130,29 +130,31 @@ foreach(GROUP BASE NOTIFICATIONS ALERTS STATELISTENER SPEECHDATA MUSICDATA CUSTO
         get_filename_component(FILE_BASE ${file} NAME_WE)
         list(APPEND PROTOC_INC_DIRS "-I.")
         list(APPEND PROTOC_INC_DIRS "-I${FILE_PATH}")
-        list(APPEND ${GROUP}_PROTO_BYPRODUCTS "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c")
+        list(APPEND ${GROUP}_PROTO_OUTPUTS "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c")
+        list(APPEND ${GROUP}_PROTO_DEPENDENCIES "${file}")
     endforeach()
 
-    # Populate list of compilation commands for protoc compiler.
-    unset(PROTOC_CMD_LIST)
+    # Each .pb.c file to be compiled by protoc gets its own CMake custom command,
+    # with dependencies to all other .proto files in its group.
     foreach(file ${${GROUP}_PROTO_FILES})
         get_filename_component(FILE_PATH ${file} DIRECTORY)
         get_filename_component(FILE_NAME ${file} NAME)
+        get_filename_component(FILE_BASE ${file} NAME_WE)
+
         set(PROTOC_ARGS "-I."
                         "${PROTOC_INC_DIRS}"
                         "--nanopb_out=${PROTO_COMPILATION_DIR}"
                         "${FILE_NAME}"
         )
-        list(APPEND PROTOC_CMD_LIST
-                    COMMAND ${CMAKE_COMMAND} -E chdir ${FILE_PATH}
-                    ${PROTOC_BIN} ${PROTOC_ARGS})
-    endforeach()
 
-    add_custom_target(
-        protobuf_generation_${GROUP}
-        ${PROTOC_CMD_LIST}
-        BYPRODUCTS ${${GROUP}_PROTO_BYPRODUCTS}
-    )
+        add_custom_command(
+          OUTPUT "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c"
+          COMMAND ${PROTOC_BIN} ${PROTOC_ARGS}
+          WORKING_DIRECTORY ${FILE_PATH}
+          MAIN_DEPENDENCY ${file}
+          DEPENDS ${${GROUP}_PROTO_DEPENDENCIES}
```

Comment:
Does compilation of one of the source depend on all of the sources files in the group?
Shouldn't this only have the MAIN_DEPENDENCY on the ${file}?

Reply:
I looked more closely at the dependencies in the .proto files.
Most of the groups have two .proto files, one of which has a dependency to the other file in its group, and additionally a dependency to `directiveHeader.proto`, which is in the `BASE` group. 

None of these .proto files are expected to change, so it might be unnecessary to have these dependencies, but to have all cases covered the `BASE` group should probably also be added as a dependency.